internal/juju/offers.go

```diff
@@ -390,3 +401,53 @@ func (c offersClient) RemoveRemoteOffer(input *RemoveRemoteOfferInput) []error {
 
 	return nil
 }
+
+// GrantOffer adds access to an offer managed by the access offer resource.
+// No action or error if the access was already granted to the user.
+func (c offersClient) GrantOffer(input *GrantRevokeOfferInput) error {
+	conn, err := c.GetConnection(nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = conn.Close() }()
+
+	client := applicationoffers.NewClient(conn)
+
+	for _, user := range input.Users {
+		err = client.GrantOffer(user, input.Access, input.OfferURL)
+		if err != nil {
+			// ignore if user was already granted
+			if strings.Contains(err.Error(), "user already has") {
+				continue
+			}
+			return err
+		}
+	}
+
+	return nil
+}
+
+// RevokeOffer revokes access to an offer managed by the access offer resource.
+// No action or error if the access was already revoked for the user.
+func (c offersClient) RevokeOffer(input *GrantRevokeOfferInput) error {
+	conn, err := c.GetConnection(nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = conn.Close() }()
+
+	client := applicationoffers.NewClient(conn)
+
+	for _, user := range input.Users {
+		err = client.RevokeOffer(user, input.Access, input.OfferURL)
+		if err != nil {
+			// ignorer if user was already revoked
+			if strings.Contains(err.Error(), "not found") {
```

Comment:
same question as above - i'm not sure what error is returned in the case where user does not have access to an offer

Reply:
I also got this while testing, I assumed that this one is from here:
https://github.com/juju/juju/blob/6e0abffcb65b54fb14cd51db007c42bc6ffb1e02/domain/access/service/permission.go#L156